File: backend/app/exceptions.py
"""Custom exception hierarchy for SecurePR AI."""


class SecurePRError(Exception):
    """Base exception for all SecurePR AI errors."""

    def __init__(self, message: str, details: dict | None = None):
        super().__init__(message)
        self.message = message
        self.details = details or {}


class ConfigurationError(SecurePRError):
    """Raised when configuration is invalid or missing."""
    pass


class LlmProviderError(SecurePRError):
    """Raised when LLM provider operation fails."""
    pass


class VCSIntegrationError(SecurePRError):
    """Raised when VCS (GitHub, GitLab, etc.) integration fails."""
    pass


class QueueError(SecurePRError):
    """Raised when queue operations fail."""
    pass


class RagError(SecurePRError):
    """Raised when RAG/vector store operations fail."""
    pass


class PipelineError(SecurePRError):
    """Raised when security analysis pipeline fails."""
    pass


class ValidationError(SecurePRError):
    """Raised when input validation fails."""
    pass


class WebhookError(SecurePRError):
    """Raised when webhook processing fails."""
    pass

File: backend/src/exceptions.ts
/**
 * Custom exception hierarchy for SecurePR AI.
 */

export class SecurePRError extends Error {
  public details: Record<string, unknown>;

  constructor(message: string, details?: Record<string, unknown>) {
    super(message);
    this.name = 'SecurePRError';
    this.details = details || {};
  }
}

export class ConfigurationError extends SecurePRError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, details);
    this.name = 'ConfigurationError';
  }
}

export class LlmProviderError extends SecurePRError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, details);
    this.name = 'LlmProviderError';
  }
}

export class VCSIntegrationError extends SecurePRError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, details);
    this.name = 'VCSIntegrationError';
  }
}

export class QueueError extends SecurePRError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, details);
    this.name = 'QueueError';
  }
}

export class RagError extends SecurePRError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, details);
    this.name = 'RagError';
  }
}

export class PipelineError extends SecurePRError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, details);
    this.name = 'PipelineError';
  }
}

export class ValidationError extends SecurePRError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, details);
    this.name = 'ValidationError';
  }
}

export class WebhookError extends SecurePRError {
  constructor(message: string, details?: Record<string, unknown>) {
    super(message, details);
    this.name = 'WebhookError';
  }
}
